### PR TITLE
feat: collector & spotter

### PR DESCRIPTION
**Pull request by manuelpolzhofer**
please review #111 first

### CODE REVIEW

File: src/core/pile.sol [hunk 1]
```diff
@@ -128,11 +128,20 @@ contract Pile is DSNote, TitleOwned {
         beans.drip(loans_[loan].fee);
```
Comment (charleenfei) [inline src/core/pile.sol]: I think we can think about taking this function abstraction out if it just calls only `beans.drip`

File: src/core/pile.sol [hunk 2]
```diff
@@ -128,11 +128,20 @@ contract Pile is DSNote, TitleOwned {
...
-        require(loans_[loan].balance == 0,"before repay loan needs to be withdrawn");
+
+    // recovery used for defaulted loans_
+    function recovery(uint loan, uint wad) public auth {
+        doRepay(loan, wad);
+
+        uint fee = loans_[loan].fee;
+        uint loss = beans.debtOf(loan, fee);
+
+        beans.decLoanDebt(loan, loans_[loan].fee, loss);
```
Comment (charleenfei) [inline src/core/pile.sol]: why use `loans_[loan].fee` here if you already declared the variable above? or maybe we don't need to declare the variable.

File: src/core/pile.sol [hunk 3]
```diff
@@ -146,6 +155,14 @@ contract Pile is DSNote, TitleOwned {
         tkn.approve(lender, wad);
     }
 
+    // repay() a certain amount of token from the user to the Pile
+    function repay(uint loan, uint wad) public owner(loan) note {
+        // moves currency from usr to pile and reduces debt
+        require(loans_[loan].balance == 0,"before repay loan needs to be withdrawn");
+        doRepay(loan, wad);
+
```
Comment (charleenfei) [inline src/core/pile.sol]: space

File: src/core/test/pile.t.sol [hunk 4]
```diff
@@ -141,4 +141,5 @@ contract PileTest is DSTest {
         uint debt = pile.burden(loan);
         repay(loan, debt);
     }
+
```
Comment (charleenfei) [inline src/core/test/pile.t.sol]: space

File: src/core/collect/collector.sol [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    function file(bytes32 what, address addr) public auth {
+        if (what == "spotter") spotter = SpotterLike(addr);
+        else if (what == "tag") tag = TagLike(addr);
+        else if (what == "desk") desk = DeskLike(desk);
+        else if (what == "pile") pile = PileLike(pile);
+        else revert();
+    }
+
+    function collect(uint loan, address usr) public auth {
+        if(spotter.collectable(loan)){
```
Comment (charleenfei) [inline src/core/collect/collector.sol]: !collectable

File: src/core/collect/spotter.sol [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.4.24;
+
+contract ShelfLike {
+    function adjust(uint loan) public;
+    function shelf(uint loan) public returns (address, uint256, uint, uint);
```
[the remark this answers is not in the document]
Comment (manuelpolzhofer) [reply src/core/collect/spotter.sol]: I think we should generate a refactoring ticket for it.

File: src/core/pile.sol (hunk 1 above)
Comment (manuelpolzhofer) [reply src/core/pile.sol]: Currently, it is easier because bean doesn't know the `fee` of a loan.

You would need to ask pile about the fee first

File: src/core/collect/collector.sol [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    function depend(bytes32 what, address addr) public auth {
+        if (what == "spotter") spotter = SpotterLike(addr);
+        else if (what == "tag") tag = TagLike(addr);
+        else if (what == "desk") desk = DeskLike(desk);
+        else if (what == "pile") pile = PileLike(pile);
+        else revert();
+    }
+
+    function collect(uint loan, address usr) public auth {
+        if(spotter.collectable(loan) == false){
```
Comment (charleenfei) [inline src/core/collect/collector.sol]: !spotter.collectable(loan)

File: src/core/collect/test/integration.t.sol [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        assertEq(pile.loan(), loan);
+    }
+
+    function testCollectWithSpotter() public {
+        uint loan = 1; uint tokenId = 123;
+        uint price = 150 ether; uint debt = 100 ether;
+        setUpLoan(loan,tokenId, price, debt);
+
+        // 150% is enough threshold is 120%
+        assertTrue(spotter.seizable(loan) == false);
```
Comment (charleenfei) [inline src/core/collect/test/integration.t.sol]: !spotter.seizable(loan)

File: src/core/collect/test/integration.t.sol [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+        // increase debt to 115%
+        pile.setDebtOfReturn(130 ether);
+        assertTrue(spotter.seizable(loan) == true);
```
Comment (charleenfei) [inline src/core/collect/test/integration.t.sol]: spotter.seizable(loan)

File: src/core/pile.sol [hunk 10]
```diff
@@ -128,11 +128,18 @@ contract Pile is DSNote, TitleOwned {
...
-    // repay() a certain amount of token from the user to the Pile
-    function repay(uint loan, uint wad) public owner(loan) note {
-        // moves currency from usr to pile and reduces debt
-        require(loans_[loan].balance == 0,"before repay loan needs to be withdrawn");
+
+    // recovery used for defaulted loans_
+    function recovery(uint loan, uint wad) public auth {
+        doRepay(loan, wad);
+
+        uint loss = beans.debtOf(loan, loans_[loan].fee);
```
Comment (charleenfei) [inline src/core/pile.sol]: this is not actually the loss but the amount we have made up from recovering the defaulted loan. the loss would be the remaining debt after the below `beans.decLoanDebt` method is called.

Comment (ilinzweilin) [reply src/core/pile.sol]: @charly doRepay is calling the beans.decLoanDebt(). From my understanding... Manuel is just using the computed loss to decrease the the debt amount in beans... because no payment for this loan is expected anymore

File: src/core/pile.sol [hunk 11]
```diff
@@ -128,11 +128,18 @@ contract Pile is DSNote, TitleOwned {
...
+        beans.decLoanDebt(loan, loans_[loan].fee, loss);
+    }
+
+    function doRepay(uint loan, uint wad) internal {
         collect(loan);
+
         uint fee = loans_[loan].fee;
         uint debt = beans.debtOf(loan, fee);
```
Comment (ilinzweilin) [inline src/core/pile.sol]: @manuelpolzhofer I think this check below prevents partial repay and consequently also recovery (which might be a partial repay)

File: src/core/collect/spotter.sol [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    function free(uint loan, address usr) public;
+}
+
+contract SPileLike {
```
Comment (ilinzweilin) [inline src/core/collect/spotter.sol]: why SPileLike and not PileLike

File: src/core/collect/spotter.sol [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    function nowPrice(uint loan) internal returns(uint) {
+        shelf.adjust(loan);
+        (,,uint price,) = shelf.shelf(loan);
+        return price;
+    }
+
+    function nowDebt(uint loan) internal returns(uint) {
+        pile.collect(loan);
```
Comment (ilinzweilin) [inline src/core/collect/spotter.sol]: could you also call: return pile.burden(loan) here?

Comment (manuelpolzhofer) [reply src/core/collect/spotter.sol]: I think we can implement it together with the cabinet contracts because it will change

File: src/core/pile.sol (hunk 10 above)
Comment (manuelpolzhofer) [reply src/core/pile.sol]: yes. I just did it in that way to re-use the existing `decLoanDebt` method to set the loan to `zero`
